### PR TITLE
plugin/template: fix CNAME upstream handling

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

Fixes a couple of bugs with template CNAME lookup handling...
1. if `upstream` is not set, we'd return a SERVFAIL if a CNAME was in the response.
2. CNAME target lookup was being performed on CNAME queries (not just A and AAAA queries).  Expected behavior is that a query for CNAME records only returns the CNAME records.  Lookup and adding of A/AAAA records should only be done for A/AAAA queries.

### 2. Which issues (if any) are related?

#1887

### 3. Which documentation changes (if any) need to be made?

none

### CODE REVIEW

**Comment by chrisohaver** [comment]
oops - i need to remove that extraneous unrelated file...

**Comment by greenpau** [comment]
@chrisohaver , any unit test improvements?

File: plugin/template/template.go [hunk 1]
```diff
@@ -84,8 +84,8 @@ func (h Handler) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 				return dns.RcodeServerFailure, err
 			}
 			msg.Answer = append(msg.Answer, rr)
-			if rr.Header().Rrtype == dns.TypeCNAME {
-				up, _ := template.upstream.Lookup(state, rr.(*dns.CNAME).Target, dns.TypeA)
+			if template.upstream != nil && (state.QType() == dns.TypeA || state.QType() == dns.TypeAAAA) && rr.Header().Rrtype == dns.TypeCNAME {
```
Comment (greenpau) [inline plugin/template/template.go:87]: @chrisohaver , please add unit tests.

**Review by greenpau — CHANGES_REQUESTED** [review]

**Comment by greenpau** [comment]
@chrisohaver , please let me know when you are done with this PR. I will approve.

**Comment by chrisohaver** [comment]
I could not work out how to test upstream function with "unit" tests, so added an integration test.

**Comment by greenpau** [comment]
> I could not work out how to test upstream function with "unit" tests, so added an integration test.

@miekg , I am also not sure about this one. The integration test is great, but unit test is better, imho. What is your take on this?

**Comment by chrisohaver** [comment]
To be clear, the integration test is in `/test`, where other similar integration tests reside.  Most of these tests run a server instance with a given corefile - hence they are technically integration tests, not strictly unit tests.

There is a similar test that tests CNAME upstream lookups for the `file` plugin (`/test/file_upstream_test.go`).

There _may_ be a way to test upstream lookups without running a server instance... but I don't see a pressing reason to do so.

**Comment by chrisohaver** [comment]
> There may be a _way_ to test upstream lookups without running a server instance.

Actually - I could do it by creating an interface for `upstream`, `upstreamer`, and mocking it out with a something that implements `upstreamer`...  this would require some refactoring in `template`.

IMO not worth the trouble unless people feel really strongly about it.

**Review by miekg — APPROVED** [review]
there is test, the only downside is that codecov doesn't count it correctly

**Comment by greenpau** [comment]
> there is test, the only downside is that codecov doesn't count it correctly

@miekg , 👍 , will remember this for the future. Thought it is better to ask.

**Comment by greenpau** [comment]
@chrisohaver , do you mind squashing? (or perhaps it is done automatically)

**Comment by miekg** [comment]
@greenpau we squash on commit (which should a project wide setting). Don't care that much on how the exact commit message looks.

**Pull request merged**